nodeJS/express/introduction_to_express.md

```diff
@@ -0,0 +1,98 @@
+
+Whenever you navigate to any web URL this way, this is essentially what you are doing. Navigating to `https://theodinproject.com/paths` via the address bar is just telling the browser to send a `GET` request to the `/paths` route at `https://theodinproject.com`, then display what it receives in response.
+
+Once you navigate to `http://localhost:3000/`, you should see `Hello, world!` appear in the window. Let's break down how our Express code got us here.
+
+When our server receives our `GET` request, Express stores the request in a [`request` object](https://expressjs.com/en/4x/api.html#req). This request gets passed through a chain of functions we call `middleware functions` until eventually, a middleware function tells Express to actually respond to the request.
+
+In our example, the request comes through as a `GET` request to the `/` route. This matches the route handler we have.
+
+```javascript
+app.get("/", (req, res) => res.send("Hello, world!"));
+```
+
+We will break down routes and route handlers in more detail in a later lesson, but to summarise the above line, it tells Express "if a `GET` request comes through to the `/` route, run the request through the following chain of middleware functions". Here, we only have a single middleware function in this chain.
+
+If we had defined multiple route handlers, Express would pass the request through the first one to match the requested route. Order matters!
+
+Express takes the callback function we gave it and passes the request object into the first parameter (commonly named `req` by convention), and a [`response` object](https://expressjs.com/en/4x/api.html#res) into the second parameter (commonly named `res`). Our callback tells the response object to respond to the request by `.send`ing the string `"Hello, world!"`.
+
+There is no more code to run and the function returns. Since Express has been told to respond to the request, it ends the request here. Meanwhile, the browser receives our server's response and displays it on screen, which is our `"Hello, world!"` string. We could send nearly anything in response. We could even [tell Express to send a file](https://expressjs.com/en/api.html#res.sendFile).
+
+### Nodemon
+
+After you run your server with `node app.js`, if you make any changes to any JavaScript and JSON files in the project directory, these changes will not be reflected in the server unless you manually interrupt and rerun `node app.js`. Similarly to how Webpack and Vite's dev servers would restart and rebuild your site when you make a change to a file, you can use [Nodemon](https://www.npmjs.com/package//nodemon) to run your `app.js` file.
+
+Our recommendation would be to add it as a dev dependency and write an npm script that runs `nodemon app.js`.
```

Comment:
Love this suggestion. Though I wonder if Node's built-in watch will suffice for the course's use case?

Reply:
I thought about that but given it's still experimental in LTS (heck, even in v21), I opted against it as I feel the curriculum should stick to recommending stable features. Similarly to how Jest's native ESM support is still experimental due to the relevant Node APIs being unstable - hence it'd be better to stick with the very quick babel setup for Jest/ESM usage.

Reply:
Makes sense